### PR TITLE
Node Cache for Builds

### PR DESCRIPTION
**Pull request by adambkaplan**
Restoring (in part) the node-level cache used in OCP 3.x builds.
This proposal utilizes buildah's additional stores feature to mount the
node image store as read-only.

[DEVEXP-334](https://issues.redhat.com/browse/DEVEXP-334)

### CODE REVIEW

**Comment by adambkaplan** [comment]
/assign @bparees 

/cc @nalind @rhatdan @TomSweeneyRedHat 

fyi @openshift/openshift-team-developer-experience

File: enhancements/builds/node-cache-for-builds.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. The build controller will mount the node image cache located at
+   `/var/lib/containers/storage` into the build pod as `readOnly`. The mount point should not
+   conflict with buildah's own default storage directory. Ex: `/var/lib/shared/containers/storage`.
+2. The openshift/builder image is modified to configure additional image stores for buildah (via
+   `/etc/containers/storage.conf`).
+3. When using the additional image store, buildah _copies_ any blobs found in the cache to its own
+   image store within the container.
```
Comment (adambkaplan) [inline enhancements/builds/node-cache-for-builds.md]: @nalind this was the biggest unknown for me when I drafted this proposal. Does buildah support this behavior currently, or does it assume that any layers in the additional store will remain there for the duration of the build?

File: enhancements/builds/node-cache-for-builds.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+1. Can cri-o be enhanced so pods are not scheduled if the kubelet/node garbage collection is in progress?
```
Comment (bparees) [inline enhancements/builds/node-cache-for-builds.md]: what's the motivation for the question?

and i think the answer is no, anyway.  crio does not control the scheduler.  at best you could have crio not *start* a container if GC was happening, maybe.

and you'd still have the problem of GC starting after the container started(which you definitely cannot block because then you'd potentially starve out GC from ever running), so i don't think it solves anything.

File: enhancements/builds/node-cache-for-builds.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+## Summary
+
+Builds in OpenShift 4.x currently require every layer of the base image to be pulled from its
+source registry. In OCP 3.x builds could take advantage of image layers that were stored on the
```
Comment (bparees) [inline enhancements/builds/node-cache-for-builds.md]:
```suggestion
source registry every time the build is run. In OCP 3.x builds could take advantage of image layers that were stored on the
```

File: enhancements/builds/node-cache-for-builds.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+build's node. This proposal provides a means of restoring this node-level cache.
```
Comment (bparees) [inline enhancements/builds/node-cache-for-builds.md]: This limitation also means we get no layer reuse when executing dockerfile commands.  Being able to use layers from disk could help us with that, though i don't think the solution we've proposed so far would actually get us that (the layers might exist on the node in the case that a previous build's output image had been pulled to the node for some reason, but buildah isn't going to see/use them during Dockerfile statement execution, i suspect.  Worth talking with @nalind about though, because being able to no-op a command like "RUN go build ." would be significant).

File: enhancements/builds/node-cache-for-builds.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+### Goals
+
+1. Provide a means for builds to take advantage of image layers that are stored on the node.
+
+### Non-Goals
+
+1. Provide a means for builds to share an image layer cache within a namespace.
+2. Provide a means for the internal registry to pre-populate an image layer cache within a
+   namespace or across a cluster.
+3. Add persistent volume support for builds.
```
Comment (bparees) [inline enhancements/builds/node-cache-for-builds.md]: provide a means for builds to write to the cache (at least in this iteration, right?)

File: enhancements/builds/node-cache-for-builds.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+1. The build controller will mount the node image cache located at
+   `/var/lib/containers/storage` into the build pod as `readOnly`. The mount point should not
+   conflict with buildah's own default storage directory. Ex: `/var/lib/shared/containers/storage`.
+2. The openshift/builder image is modified to configure additional image stores for buildah (via
+   `/etc/containers/storage.conf`).
+3. When using the additional image store, buildah _copies_ any blobs found in the cache to its own
+   image store within the container.
+
+### Risks and Mitigations
```
Comment (bparees) [inline enhancements/builds/node-cache-for-builds.md]: additional risk:  unbound image cache growth.  mitigation: node cache is read-only(so builds don't contribute to it, and GC continues to manage it), builds make a temporary copy of the cached content they need, temporary copy is deleted at the end of the build.

File: enhancements/builds/node-cache-for-builds.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+## Drawbacks
+
+1. Mounting content from the node requires the build pod to continue running as privileged.
+2. No API changes proposed - there is no means to opt-out of this feature.
+
+## Alternatives
+
+1. Use persistent volumes to cache layers for builds
+   1. General build volume support is being considered as a separate enhancement.
+   2. Automatic provisioning of block storage is not guaranteed in all environments.
```
Comment (bparees) [inline enhancements/builds/node-cache-for-builds.md]: i think this also has challenges in terms of write-atomicity and performance.

plus requiring something additional to GC this new cache.

File: enhancements/builds/node-cache-for-builds.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+2. Mount the node container storage directly
+   1. Instead of using additional stores, mount the node cache as buildah's container storage.
+   2. This is very insecure, as the volume mount would need to be writeable.
```
Comment (bparees) [inline enhancements/builds/node-cache-for-builds.md]: and conflicts with node GC (node GC would be unaware that anything was consuming the image content we were putting in the cache and could delete it out from under us)

File: enhancements/builds/node-cache-for-builds.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+## Implementation History
+
+2020-02-18: Initial draft
+
+## Drawbacks
+
+1. Mounting content from the node requires the build pod to continue running as privileged.
+2. No API changes proposed - there is no means to opt-out of this feature.
+
+## Alternatives
```
Comment (bparees) [inline enhancements/builds/node-cache-for-builds.md]: i think dan walsh had some more alternatives in his doc, outlining the spectrum of how much image cache sharing we wanted to do (e.g. sharing within a namespace). all of those solutions require us to manage our own cache though, so we can share it appropriately and that means we need to bound its size somehow (GC it) as well as figure out where it lives (on the node, in a PV, etc).

File: enhancements/builds/node-cache-for-builds.md (hunk 1 above)
Comment (nalind) [reply enhancements/builds/node-cache-for-builds.md]: Buildah does not currently know how to copy layer contents from read-only storage (such as CRI-O's, if it were shared) into its own read-write area.  A spot check pointed to problems that the storage library has to solve to be able to extract the contents of layers from read-only storage.  This would be necessary in order to be able to push the contents of those layers to a registry, so we need to solve those either way.

File: enhancements/builds/node-cache-for-builds.md (hunk 2 above)
Comment (derekwaynecarr) [reply enhancements/builds/node-cache-for-builds.md]: No.  Kubelet directs CRI-O to do GC.

File: enhancements/builds/node-cache-for-builds.md (hunk 6 above)
Comment (derekwaynecarr) [reply enhancements/builds/node-cache-for-builds.md]: is there any issue ensuring that the build pod has access to the layer?

File: enhancements/builds/node-cache-for-builds.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+3. When using the additional image store, buildah _copies_ any blobs found in the cache to its own
+   image store within the container.
+
+### Risks and Mitigations
+
+#### Image layer pruned by kubelet GC
+
+*Risk:* Kubelet removing image layer via GC process
+
+*Mitigation:* Buildah will be responsible for copying layers found in the additional image store
```
Comment (derekwaynecarr) [inline enhancements/builds/node-cache-for-builds.md]: how expensive is the copy?

Comment (derekwaynecarr) [reply enhancements/builds/node-cache-for-builds.md]: i know we do not make an ephemeral storage request for build pods today, but its worth noting in the future how that may or may not be improved.

see: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#requests-and-limits-setting-for-local-ephemeral-storage

right now, builds will continue to have best effort access to disk.

File: enhancements/builds/node-cache-for-builds.md (hunk 6 above)
Comment (bparees) [reply enhancements/builds/node-cache-for-builds.md]: the layer that exists on the node?  we're hostpath mounting it, i guess we'll need to make sure all the permissions are right but there shouldn't be a technical obstacle.  (whether it means giving the build pod more permissions than we'd like to is another question, though)

File: enhancements/builds/node-cache-for-builds.md (hunk 5 above)
Comment (mrunalp) [reply enhancements/builds/node-cache-for-builds.md]: What will manage the cache?

Comment (bparees) [reply enhancements/builds/node-cache-for-builds.md]: the cache gets populated by the node when it pulls images for pods.

the build just benefits from any images that happen to be in that node cache.

File: enhancements/builds/node-cache-for-builds.md (hunk 2 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: I recall this coming up during our conversation with @mrunalp, @nalind, and @rhatdan back in December 2019. It would be nice if build pods didn't start while GC is in progress, but I agree that we still run the risk of caches being pruned during the build.

I can strike this as an open question.

File: enhancements/builds/node-cache-for-builds.md (hunk 10 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: @derekwaynecarr we don't make those requests by default, but `Build`/`BuildConfig` does have a `Resources` portion of its spec which does eventually make it to the pod template. This is something we may need to document if customers are using "size of source repo" to ballpark how much ephemeral storage they need.

File: enhancements/builds/node-cache-for-builds.md (hunk 9 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: Added a "shared RWX" cache as an alternative, the hypothetical being S3 as the storage backend.

File: enhancements/builds/node-cache-for-builds.md (hunk 7 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: Noted

File: enhancements/builds/node-cache-for-builds.md (hunk 5 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: Added as the first non-goal

File: enhancements/builds/node-cache-for-builds.md (hunk 4 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: Added a sentence here denoting this limitation.

File: enhancements/builds/node-cache-for-builds.md (hunk 1 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: Added the `containers/storage` enhancement to the implementation details.

File: enhancements/builds/node-cache-for-builds.md [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+### User Story
+
+As a developer using OpenShift to build container images I want builds to use image layers cached on
+the node So that builds don't have to pull all layers in their base images And builds run faster.
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. The build controller will mount the node image cache located at `/var/lib/containers/storage`
+   into the build pod as `readOnly`. The mount point should not conflict with buildah's own default
+   storage directory. Ex: `/var/lib/shared/containers/storage`.
```
Comment (gabemontero) [inline enhancements/builds/node-cache-for-builds.md]: @adambkaplan - am I correct in thinking that if  in tekton/buildv2, where the `buildah` cli is directly invoked in one tekton `corev.Container` 's (the tekton `Step`) that use the `buildah` image, then the user simply needs to set the appropriate `buildah bud` option (perhaps `--cache-from` ? ... or is that a question for @nalind or @TomSweeneyRedHat  ?)

Anyway, understanding the CLI consumption mechanism in addition to the API ones done in openshift/builder is an important end goal here, IMO, as we discussed in planning as part of facilitating tekton/buildv2

File: enhancements/builds/node-cache-for-builds.md [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+2. The openshift/builder image is modified to configure additional image stores for buildah (via
+   `/etc/containers/storage.conf`).
+3. `containers/storage` is enhanced to copy blobs found in the additional image stores to the main
+   read-write storage (used by buildah, podman, etc.).
```
Comment (gabemontero) [inline enhancements/builds/node-cache-for-builds.md]: So if containres/storage needs updating, that means we will need an updated `buildah` binary at some point for the equivalent function, right @adambkaplan @nalind @TomSweeneyRedHat ?

File: enhancements/builds/node-cache-for-builds.md [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+As a developer using OpenShift to build container images I want builds to use image layers cached on
+the node So that builds don't have to pull all layers in their base images And builds run faster.
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. The build controller will mount the node image cache located at `/var/lib/containers/storage`
+   into the build pod as `readOnly`. The mount point should not conflict with buildah's own default
+   storage directory. Ex: `/var/lib/shared/containers/storage`.
+2. The openshift/builder image is modified to configure additional image stores for buildah (via
+   `/etc/containers/storage.conf`).
```
Comment (gabemontero) [inline enhancements/builds/node-cache-for-builds.md]: To echo the planning discussion again around enabling as much as we can tekton/buildv2 with all this, if the processing of `/etc/containers/storage.conf` can be done in a new utility repo that can be imported by openshift/builder and any tekton utilities / build v2 operators, etc. (vs. those operators / utilities having to import openshift/builder for any common code), that would be ideal.

When constructing the common code, some sort of abstraction layer to deal with seeding buildah CLI vs. API in the different targets is warranted.

Comment (gabemontero) [reply enhancements/builds/node-cache-for-builds.md]: FWIW I've called out a "common utility repo" in the entitlements proposal as well.

File: enhancements/builds/node-cache-for-builds.md (hunk 12 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: @gabemontero correct - changes to containers/storage will need to be added to buildah.

In openshift/builder we rarely bump containers/storage independently of buildah - and in this case we would likely need to bump the two in sync.

Comment (gabemontero) [reply enhancements/builds/node-cache-for-builds.md]: Thanks for the confirmation @adambkaplan 

Should we cite this then as an explicit dependency on the containers team?  Or do you envision whoever from our team implements this also having to make the buildah changes, prior to the openshift/builder go mod bump?

File: enhancements/builds/node-cache-for-builds.md (hunk 13 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: @gabemontero So today we bake in a `storage.conf` file into the openshift/builder image, and right now we don't modify this file via the build controller [1]. The additional stores are in fact already configured - we may not need to do anything to this file as a part of the enhancement [2].

For Tekton/builds v2, the images which contain buildah will need to likewise be updated to have a default `storage.conf` that could take advantage of the additional store (with the appropriate mounts and buildah CLI options).

[1] https://github.com/openshift/openshift-controller-manager/blob/master/pkg/build/controller/strategy/util.go#L382-L411
[2] https://github.com/openshift/builder/blob/master/imagecontent/storage.conf#L19-L23

Comment (gabemontero) [reply enhancements/builds/node-cache-for-builds.md]: Makes sense @adambkaplan ... and I see you've pushed a commit that calls out the need for both buildah CLI and API updates to leverage this.  LGTM.

I'll ping the app svcs/dev tools tekton guys about this in slack.  Certainly feel free to tag them in this review as you see fit.

**Comment by bparees** [comment]
needs containers team signoff but looks good from my perspective.

File: enhancements/builds/node-cache-for-builds.md (hunk 13 above)
Comment (gabemontero) [reply enhancements/builds/node-cache-for-builds.md]: makes sense @bparees 

I think details / lessons learned from this work, plus ensuring any buildah API enhancements around specifying storage are available from the CLI (and on the suface I don't see why they would do it just for API vs. CLI), would be a sufficient starting point for tekton based solutions built by our brethren on that side of the shop.

File: enhancements/builds/node-cache-for-builds.md [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+that need to be pulled when the build commences.
+
+### User Story
+
+As a developer using OpenShift to build container images I want builds to use image layers cached on
+the node So that builds don't have to pull all layers in their base images And builds run faster.
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. The build controller will mount the node image cache located at `/var/lib/containers/storage`
```
Comment (ricardomaraschini) [inline enhancements/builds/node-cache-for-builds.md]: I have a question about this: If I assume that the images used during a build process  are not images the cluster generally runs as pod when are we going to have a hit on the cache? 

e.g. if I use a golang image during a build this image wont be cached(as the cache is readonly) and as i never use to run this very same golang image in a Deployment/Pod it will never be cached.

Comment (bparees) [reply enhancements/builds/node-cache-for-builds.md]: well if you run an image that's based on that golang image, the base layer would still get into the cache.

but yes if you use it as a base for a multistage build you're correct.  or if you just build the image and push it but never deploy it.  (adam has some references to these scenarios elsewhere in the doc)

that said, in many cases the the build is based on "ubi" or "centos" in which case we would still get benefit.

Comment (bparees) [reply enhancements/builds/node-cache-for-builds.md]: (or even a ruby or nodejs base image, which when the app image that's built on top of it gets deployed, would result in the ruby/nodejs base getting pulled into the cache)

Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: @ricardomaraschini I call this out in the drawbacks section. S2I builds will likely benefit the most because these tend to produce "fat" images that are loaded with test/dev tools. Multistage Dockerfile builds will likely benefit the least, since developers often use these to produce "skinny"/"runtime" images.

File: enhancements/builds/node-cache-for-builds.md (hunk 11 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: From @rhatdan's blog post [1], this isn't a CLI option, but a config option within `storage.conf`.

[1] https://developers.redhat.com/blog/2019/08/14/best-practices-for-running-buildah-in-a-container/

**Comment by adambkaplan** [comment]
/cc @mrunalp @nalind @runcom 

Pinging folks from the containers-engine team on this thread. It seems that we may need to enhance the `containers/storage` library (or buildah) such that cache hits in the read-only additional store can be copied into buildah's read/writeable store. Otherwise we risk builds failing due to layers being removed from underneath the build container.

**Comment by cgwalters** [comment]
This will also involve potentially different versions of the containers/storage library concurrently accessing the same storage, right?  Seems safer to have a high level protocol where the build pod can basically:

- list images on the host
- ask a process running on the host to serialize an image to a pipe/socket, then deserialize

That also avoids issues with locking.

If we wanted to have the build pod directly use the merged (mounted) filesystem path, we could potentially extend the protocol to have the build pod request locks on images via the protocol.

Basically I'm arguing for having two processes communicate over e.g. a local Unix domain socket - if we wanted to pass merged/mounted filesystem trees, that could be done via file descriptor passing of a directory FD or so?

**Comment by bparees** [comment]
> This will also involve potentially different versions of the containers/storage library concurrently accessing the same storage

only to read the layers.  Is there a versioning issue there?

**Comment by cgwalters** [comment]
> only to read the layers. Is there a versioning issue there?

I'm not a containers/storage maintainer; but today e.g. for OpenShift we have a `crio-wipe.service` which blows away the container storage across major version upgrades because they didn't want to have to think about upgrade compatibility.  podman doesn't do that - there have been bugs (though I'm not sure if those are in the podman layer or lower)

Here we're not just talking about handling the case of N+1 being able to read N - the host container storage version may be either older or newer than the one in the build container in the general case.

**Comment by adambkaplan** [comment]
@cgwalters can `containers/storage` bump its major version as part of an OpenShift z-stream update? Requiring buildah, openshift builds, and crio to be in sync with respect to the major version of `containers/storage` is a drawback, but not insurmountable IMO.

**Comment by bparees** [comment]
Not clear to me the implications for consuming existing storage(created
before upgrade) after an upgrade.


Ben Parees | OpenShift

**Comment by adambkaplan** [comment]
Per discussion with @nalind version skew with `containers/storage` is a low probability issue - the actual storage format is considered stable/backwards compatible between major versions (breaking this is considered a bug).

We do have a separate blocker in that today buildah/podman can't push layers that are present in the additional storage layer. This is problematic when pushing to the internal registry - there is no guarantee that layers cached on the node are present in the internal registry (or any external registry for that matter).

**Comment by adambkaplan** [comment]
@cgwalters is the version of `containers/storage` something that would have a major version bump delivered as a z-stream OpenShift upgrade? Ex: If we upgrade from 4.5.0 to 4.5.1, could the version of `containers/storage` change from v5 to v6?

File: enhancements/builds/node-cache-for-builds.md (hunk 8 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: Noted that the GC issue impacts layers read _and_ written.

File: enhancements/builds/node-cache-for-builds.md (hunk 6 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: @derekwaynecarr @bparees is this something that we address separately in a proof of concept or during implementation?

File: enhancements/builds/node-cache-for-builds.md [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+
+#### Containers/storage upgrades can make builds unusable
+
+*Risk:* On upgrade, version skew between the `containers/storage` library used by cri-o and buildah
+can cause builds to be unusable.
+
+*Mitigation:* The image layer format in use by `containers/storage` is designed to be backwards
+compatible between major versions. If in the event data in the additional store is not usable,
+buildah should fall back to its default behavior of pulling image content from the upstream
+registry.
```
Comment (adambkaplan) [inline enhancements/builds/node-cache-for-builds.md]: @nalind does buildah behave like this today?

File: enhancements/builds/node-cache-for-builds.md (hunk 6 above)
Comment (bparees) [reply enhancements/builds/node-cache-for-builds.md]: it's not clear to me that there's anything to implement for either of the issues raised.  So i would just expect the EP to list the issue, describe why it won't be a problem given our architecture, and move on.

Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: Added mitigations below:

1. Host path doesn't exist (create it)
2. Host path does not allow privileged containers to read (worst case fall back to current behavior)

File: enhancements/builds/node-cache-for-builds.md [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+behavior if the additional storage directory is empty.
+
+#### Build container does not have permission to read the image cache
+
+*Risk:* Build container does not have permission to read the image cache on the node.
+
+*Mitigation:* Builds already run as a privileged container, and thus are able to obtain more
+privileges than the parent kubelet. If for any reason the node does not allow privileged containers
+to read contents in the node image cache, buildah should be able to fall back to its current
+behavior.
```
Comment (adambkaplan) [inline enhancements/builds/node-cache-for-builds.md]: @nalind another item that we should verify - if the additional store can't be read by buildah, will it default to "no additional store?"

File: enhancements/builds/node-cache-for-builds.md (hunk 12 above)
Comment (adambkaplan) [reply enhancements/builds/node-cache-for-builds.md]: @gabemontero per discussion with @nalind we will need an enhancement from buildah to be eventually vendored in. That shouldn't stop us from doing a proof of concept - some of the code already exists in openshift-controller-manager.

Comment (gabemontero) [reply enhancements/builds/node-cache-for-builds.md]: Sounds good @adambkaplan .... and from taking a quick peek at the proposal since I originally posted https://github.com/openshift/enhancements/pull/216#discussion_r383918898 there has been some evolution with the proposal, and I see some explicit call outs to buildah/storage changes, etc., so I would agree the dependency we have has been sufficiently identified.

thanks

**Comment by cgwalters** [comment]
> @cgwalters is the version of containers/storage something that would have a major version bump delivered as a z-stream OpenShift upgrade? Ex: If we upgrade from 4.5.0 to 4.5.1, could the version of containers/storage change from v5 to v6?

Remember, containers/storage doesn't exist on its own - it's a Go library that gets vendored into another binary, most notably `podman` and `crio` (currently, independently).

The answer overall then is that RHEL can absolutely rebase mid-OpenShift Z, and that has already happened for 8.0 :arrow_right: 8.1 in 4.2.  Now, *almost* the entire platform is structured as pods scheduled by kubelet.  The main exceptions are the bootstrap process and [OS updates](https://github.com/openshift/machine-config-operator/blob/master/docs/OSUpgrades.md), both of which use `podman`.  Of those two, we obviously only care about in-place changes for the latter.

Now I forget offhand if crio and podman share an image store or not.  That's an important thing to determine.  If they do we already had two c/storage versions in play already, having build pods access that same storage would bring a potential third version in.

**Comment by bparees** [comment]
@adambkaplan it lgtm, ping me tomorrow if you're ready for the label (allowing some time for others to make a final pass if they desire)

**Comment by rhatdan** [comment]
podman and CRI-O share the image store.

**Comment by adambkaplan** [comment]
@bparees @nalind @cgwalters added a note on the existing containers/storage version skew we have with podman and cri-o. Squashed commits and marked `implementable`.
